app/src/chat_engine.py

```diff
@@ -85,18 +85,34 @@ class BaseEngine(ChatEngineInterface):
     ]
 
     def on_message(self, question: str, chat_history: list[dict[str, str]]) -> OnMessageResult:
-        chunks_with_scores = retrieve_with_scores(
+        attributes = analyze_message(self.llm, question)
+        if attributes.needs_context:
+            question = question if attributes.is_in_english else attributes.message_in_english
+
+            chunks_with_scores = retrieve_with_scores(
+                question,
+                retrieval_k=self.retrieval_k,
+                retrieval_k_min_score=self.retrieval_k_min_score,
+                datasets=self.datasets,
+            )
+
+            chunks = [chunk_with_score.chunk for chunk_with_score in chunks_with_scores]
+            # Provide a factory to reset the citation id counter
+            subsections = split_into_subsections(chunks, factory=CitationFactory())
+            context_text = create_prompt_context(subsections)
+        else:
+            context_text = None
+            chunks_with_scores = []
+            subsections = []
```

Comment:
`context_text` doesn't need to be set.
Consider:
```suggestion
            return OnMessageResult(response, self.system_prompt)
```
and setting defaults on `OnMessageResult` for `chunks_with_scores` and `subsections`.

Reply:
This will also make it more obvious the 2 scenarios where `OnMessageResult` is created.

Reply:
👍  Made this change and split the two out into separate functions as well